File: Server/src/services/tools/play_smoke_test.py
"""Play-mode smoke test: enter play, run N seconds, collect errors, exit, verdict.

Closes the agentic run-and-verify loop that run_tests does not: run_tests only
covers the test framework, not "does the game actually boot when you press Play".

Entering play mode triggers a Unity domain reload, so the C# side cannot await a
single call across play entry — it is a job (start + poll). This module exposes:

  - action="start"  -> validate preconditions and enter play mode, return job_id
  - action="status" -> poll the persisted job's state/result
  - action="run"    -> convenience composition: start, then poll with backoff
                       until done or an overall timeout, returning a compact
                       verdict. Prefer this for the common case.
"""
from __future__ import annotations

import asyncio
from typing import Annotated, Any, Literal

from fastmcp import Context
from mcp.types import ToolAnnotations

from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
import transport.unity_transport as unity_transport
from transport.legacy.unity_connection import async_send_command_with_retry

ALL_ACTIONS = ["run", "start", "status"]

DEFAULT_DURATION_SECONDS = 15
MAX_DURATION_SECONDS = 120
MAX_SAMPLE_ERRORS = 10


async def _send(ctx: Context, params: dict[str, Any]) -> dict[str, Any]:
    unity_instance = await get_unity_instance_from_context(ctx)
    result = await unity_transport.send_with_unity_instance(
        async_send_command_with_retry, unity_instance, "play_smoke_test", params
    )
    return result if isinstance(result, dict) else {"success": False, "error": str(result)}


def _clamp_duration(seconds: int) -> int:
    if seconds <= 0:
        return DEFAULT_DURATION_SECONDS
    return min(seconds, MAX_DURATION_SECONDS)


@mcp_for_unity_tool(
    group="testing",
    description=(
        "Boot-check the game in play mode in ONE call. Enters play mode, runs for "
        "duration_seconds, collects console errors/exceptions raised while playing, "
        "exits play mode, and returns a pass/fail verdict. This is the run-and-verify "
        "step run_tests does NOT cover (run_tests only runs the test framework).\n"
        "Actions:\n"
        "- run (default): start + poll until done, returns {completed, error_count, "
        "exception_count, warning_count, sample_errors, verdict, reason}. Use this.\n"
        "- start: enter play mode and return a job_id immediately (advanced).\n"
        "- status: poll a job_id started by 'start' (advanced).\n"
        "verdict is 'fail' if any errors/exceptions were raised or the job failed to "
        "boot (e.g. a compile error prevented play mode)."
    ),
    annotations=ToolAnnotations(
        title="Play Smoke Test",
    ),
)
async def play_smoke_test(
    ctx: Context,
    action: Annotated[
        Literal["run", "start", "status"],
        "run (default, one-call verdict), start (return job_id), or status (poll).",
    ] = "run",
    seconds: Annotated[
        int | str,
        "How long to stay in play mode (default 15, max 120).",
    ] | None = None,
    job_id: Annotated[str, "Job id to poll (required for action='status')."] | None = None,
    timeout_seconds: Annotated[
        int,
        "For action='run': max seconds to wait for the whole run before giving up.",
    ] | None = None,
) -> dict[str, Any]:
    action_lower = action.lower() if isinstance(action, str) else action
    if action_lower not in ALL_ACTIONS:
        return {
            "success": False,
            "error": f"Unknown action '{action}'. Valid actions: {', '.join(ALL_ACTIONS)}",
        }

    try:
        duration = int(seconds) if seconds is not None else DEFAULT_DURATION_SECONDS
    except (TypeError, ValueError):
        return {"success": False, "error": "seconds must be an integer"}
    duration = _clamp_duration(duration)

    if action_lower == "start":
        return await _send(ctx, {"action": "start", "duration_seconds": duration})

    if action_lower == "status":
        if not job_id:
            return {"success": False, "error": "action='status' requires job_id"}
        return await _send(ctx, {"action": "status", "job_id": job_id})

    # action == "run": compose start -> poll status -> verdict.
    return await _run_and_verify(ctx, duration, timeout_seconds)


async def _run_and_verify(
    ctx: Context,
    duration: int,
    timeout_seconds: int | None,
) -> dict[str, Any]:
    # Overall budget: the play duration plus generous headroom for the domain
    # reload on play entry/exit and finalization.
    if timeout_seconds is None or timeout_seconds <= 0:
        overall_timeout = duration + 90
    else:
        overall_timeout = int(timeout_seconds)

    start = await _send(ctx, {"action": "start", "duration_seconds": duration})
    if not start.get("success", False):
        return {
            "success": False,
            "completed": False,
            "verdict": "fail",
            "reason": start.get("error") or start.get("message") or "failed to start",
            "error": start.get("error"),
        }

    job_id = (start.get("data") or {}).get("job_id")
    if not job_id:
        return {
            "success": False,
            "completed": False,
            "verdict": "fail",
            "reason": "start did not return a job_id",
        }

    loop = asyncio.get_event_loop()
    deadline = loop.time() + overall_timeout
    # Backoff schedule (seconds) between status polls. Play entry triggers a domain
    # reload, so early polls may transiently fail — that is tolerated below.
    delay = 1.0
    max_delay = 4.0

    last_data: dict[str, Any] = {}
    while True:
        remaining = deadline - loop.time()
        if remaining <= 0:
            return _verdict_from(
                job_id, last_data, duration,
                completed=False,
                timed_out=True,
            )

        await asyncio.sleep(min(delay, max(0.1, remaining)))
        delay = min(max_delay, delay * 1.5)

        status = await _send(ctx, {"action": "status", "job_id": job_id})
        if not status.get("success", False):
            # During the play-entry domain reload the bridge can briefly be
            # unavailable; keep polling until the deadline instead of failing.
            continue

        last_data = status.get("data") or {}
        job_status = str(last_data.get("status", "")).lower()
        if job_status in ("done", "failed"):
            return _verdict_from(
                job_id, last_data, duration,
                completed=True,
                timed_out=False,
            )


def _verdict_from(
    job_id: str,
    data: dict[str, Any],
    duration: int,
    *,
    completed: bool,
    timed_out: bool,
) -> dict[str, Any]:
    error_count = int(data.get("error_count", 0) or 0)
    exception_count = int(data.get("exception_count", 0) or 0)
    warning_count = int(data.get("warning_count", 0) or 0)
    sample_errors = list(data.get("sample_errors") or [])[:MAX_SAMPLE_ERRORS]
    job_status = str(data.get("status", "")).lower()
    reason = data.get("reason")

    failed = (
        job_status == "failed"
        or error_count > 0
        or exception_count > 0
        or not completed
    )
    verdict = "fail" if failed else "pass"

    if reason is None:
        if timed_out:
            reason = f"timed out waiting for smoke test (job {job_id})"
        elif verdict == "fail":
            reason = "errors or exceptions were raised during play"
        else:
            reason = "no errors or exceptions during play"

    return {
        "success": True,
        "job_id": job_id,
        "completed": completed,
        "duration_seconds": duration,
        "error_count": error_count,
        "exception_count": exception_count,
        "warning_count": warning_count,
        "sample_errors": sample_errors,
        "verdict": verdict,
        "reason": reason,
    }

File: MCPForUnity/Editor/Tools/PlaySmokeTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using MCPForUnity.Editor.Helpers;
using MCPForUnity.Editor.Services;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using UnityEditor;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// One-call "does the game actually boot" smoke test: enter play mode, run for
    /// a fixed duration, collect console errors/exceptions raised while playing,
    /// exit play mode, and report a structured verdict.
    ///
    /// Entering play mode triggers a domain reload by default, which destroys all
    /// in-memory state — so this CANNOT be a single awaited handler across play
    /// entry. Instead it is a job: <c>start</c> validates preconditions, records a
    /// baseline, persists the job to <see cref="SessionState"/> (which survives the
    /// reload), and enters play mode. An <see cref="PlaySmokeTestReloadHandler"/>
    /// re-arms an <c>EditorApplication.update</c> tick after every reload that ends
    /// play mode when the duration elapses and finalizes the result on return to
    /// edit mode. <c>status</c> polls the persisted job.
    /// </summary>
    [McpForUnityTool("play_smoke_test", AutoRegister = false, Group = "testing")]
    public static class PlaySmokeTest
    {
        public static object HandleCommand(JObject @params)
        {
            var p = new ToolParams(@params);
            string action = p.Get("action", "start").ToLowerInvariant();

            switch (action)
            {
                case "start":
                    return Start(p);
                case "status":
                    return Status(p);
                default:
                    return new ErrorResponse(
                        $"Unknown action: '{action}'. Valid actions are 'start' or 'status'.");
            }
        }

        private static object Start(ToolParams p)
        {
            // Preconditions: never start on top of a compile, an active play session,
            // an in-flight smoke job, or a running test run.
            if (EditorApplication.isCompiling || EditorApplication.isUpdating)
            {
                return new ErrorResponse("editor_busy", new
                {
                    reason = "editor_busy",
                    retry_after_ms = 3000,
                });
            }

            if (EditorApplication.isPlayingOrWillChangePlaymode)
            {
                return new ErrorResponse("already_playing", new { reason = "already_playing" });
            }

            if (TestRunStatus.IsRunning)
            {
                return new ErrorResponse("tests_running", new
                {
                    reason = "tests_running",
                    retry_after_ms = 5000,
                });
            }

            if (PlaySmokeJobManager.HasActiveJob)
            {
                return new ErrorResponse("smoke_test_running", new
                {
                    reason = "smoke_test_running",
                    job_id = PlaySmokeJobManager.CurrentJobId,
                });
            }

            int durationSeconds = p.GetInt("duration_seconds", PlaySmokeJobManager.DefaultDurationSeconds)
                ?? PlaySmokeJobManager.DefaultDurationSeconds;

            string jobId = PlaySmokeJobManager.StartJob(durationSeconds);
            if (jobId == null)
            {
                return new ErrorResponse("smoke_test_running", new { reason = "smoke_test_running" });
            }

            return new SuccessResponse("Play smoke test started.", new
            {
                job_id = jobId,
                status = "running",
                duration_seconds = PlaySmokeJobManager.ClampDuration(durationSeconds),
            });
        }

        private static object Status(ToolParams p)
        {
            string jobId = p.Get("job_id");
            if (string.IsNullOrWhiteSpace(jobId))
            {
                return new ErrorResponse("Missing required parameter 'job_id'.");
            }

            var payload = PlaySmokeJobManager.GetSerializableJob(jobId);
            if (payload == null)
            {
                return new ErrorResponse("Unknown job_id.");
            }

            return new SuccessResponse("Play smoke test status retrieved.", payload);
        }
    }

    /// <summary>
    /// Persists the single active smoke-test job across domain reloads via
    /// <see cref="SessionState"/> and evaluates play/exit/finalize transitions.
    /// </summary>
    internal static class PlaySmokeJobManager
    {
        public const int DefaultDurationSeconds = 15;
        public const int MaxDurationSeconds = 120;

        // Grace period for play mode to actually begin after we request it. If the
        // project has a compile error, play mode never enters — fail after this.
        private const long EnterGraceMs = 30_000;
        private const int MaxSampleErrors = 10;

        private const string SessionKeyJob = "MCPForUnity.PlaySmokeJobV1";

        private static readonly object LockObj = new();

        internal enum Phase
        {
            AwaitingPlay,   // start requested play mode; waiting for EnteredPlayMode
            Playing,        // in play mode; counting down duration
            Finalizing,     // duration elapsed; ExitPlaymode requested
            Done,
            Failed
        }

        [Serializable]
        internal sealed class Job
        {
            public string job_id;
            public string phase;
            public long started_utc_ms;
            public long? play_entered_utc_ms;
            public long? finished_utc_ms;
            public int duration_seconds;

            // Console baseline captured at start AND re-captured right after play
            // entry (Clear on Play wipes it, so the post-entry baseline is what we
            // diff against for the verdict).
            public int baseline_errors;
            public int baseline_warnings;
            public int baseline_logs;
            public bool rebaselined;

            public int error_count;
            public int exception_count;
            public int warning_count;
            public string reason;
            public List<string> sample_errors;
        }

        public static int ClampDuration(int durationSeconds)
        {
            if (durationSeconds <= 0) return DefaultDurationSeconds;
            return Math.Min(durationSeconds, MaxDurationSeconds);
        }

        public static bool HasActiveJob
        {
            get
            {
                var job = Load();
                return job != null && (job.phase == Phase.AwaitingPlay.ToString()
                                       || job.phase == Phase.Playing.ToString()
                                       || job.phase == Phase.Finalizing.ToString());
            }
        }

        public static string CurrentJobId => Load()?.job_id;

        public static string StartJob(int durationSeconds)
        {
            lock (LockObj)
            {
                var existing = Load();
                if (existing != null && (existing.phase == Phase.AwaitingPlay.ToString()
                                         || existing.phase == Phase.Playing.ToString()
                                         || existing.phase == Phase.Finalizing.ToString()))
                {
                    return null;
                }

                GetCounts(out int e, out int w, out int l);

                var job = new Job
                {
                    job_id = Guid.NewGuid().ToString("N"),
                    phase = Phase.AwaitingPlay.ToString(),
                    started_utc_ms = NowMs(),
                    play_entered_utc_ms = null,
                    finished_utc_ms = null,
                    duration_seconds = ClampDuration(durationSeconds),
                    baseline_errors = e,
                    baseline_warnings = w,
                    baseline_logs = l,
                    rebaselined = false,
                    error_count = 0,
                    exception_count = 0,
                    warning_count = 0,
                    reason = null,
                    sample_errors = new List<string>(),
                };
                Save(job);

                // Request play mode. The subsequent domain reload is expected; the
                // reload handler re-arms the tick that drives the rest of the job.
                try
                {
                    EditorApplication.EnterPlaymode();
                }
                catch (Exception ex)
                {
                    job.phase = Phase.Failed.ToString();
                    job.reason = $"failed_to_enter_play: {ex.Message}";
                    job.finished_utc_ms = NowMs();
                    Save(job);
                }

                return job.job_id;
            }
        }

        public static object GetSerializableJob(string jobId)
        {
            var job = Load();
            if (job == null || !string.Equals(job.job_id, jobId, StringComparison.Ordinal))
            {
                return null;
            }

            string status = job.phase switch
            {
                nameof(Phase.Done) => "done",
                nameof(Phase.Failed) => "failed",
                _ => "running",
            };

            return new
            {
                job_id = job.job_id,
                status = status,
                phase = job.phase,
                duration_seconds = job.duration_seconds,
                started_utc_ms = job.started_utc_ms,
                finished_utc_ms = job.finished_utc_ms,
                error_count = job.error_count,
                exception_count = job.exception_count,
                warning_count = job.warning_count,
                reason = job.reason,
                sample_errors = (job.sample_errors ?? new List<string>()).ToArray(),
            };
        }

        // --- Reload-driven lifecycle (called by PlaySmokeTestReloadHandler) ---

        /// <summary>Called on EnteredPlayMode: re-baseline (Clear on Play wiped the console).</summary>
        public static void OnEnteredPlayMode()
        {
            lock (LockObj)
            {
                var job = Load();
                if (job == null || job.phase != Phase.AwaitingPlay.ToString())
                {
                    return;
                }

                GetCounts(out int e, out int w, out int l);
                job.baseline_errors = e;
                job.baseline_warnings = w;
                job.baseline_logs = l;
                job.rebaselined = true;
                job.play_entered_utc_ms = NowMs();
                job.phase = Phase.Playing.ToString();
                Save(job);
            }
        }

        /// <summary>Called on EnteredEditMode: finalize an active job (collect verdict).</summary>
        public static void OnEnteredEditMode()
        {
            lock (LockObj)
            {
                var job = Load();
                if (job == null)
                {
                    return;
                }

                if (job.phase == Phase.Done.ToString() || job.phase == Phase.Failed.ToString())
                {
                    return;
                }

                Finalize(job, failedReason: null);
            }
        }

        /// <summary>
        /// Per-frame tick (armed after every reload). Drives the countdown and the
        /// enter-grace timeout. Runs on the main thread.
        /// </summary>
        public static void Tick()
        {
            Job job;
            lock (LockObj)
            {
                job = Load();
                if (job == null)
                {
                    return;
                }

                if (job.phase == Phase.Done.ToString() || job.phase == Phase.Failed.ToString())
                {
                    return;
                }

                long now = NowMs();

                if (job.phase == Phase.AwaitingPlay.ToString())
                {
                    // Play mode may not have fired the state-changed callback yet in
                    // this reload cycle; if we are actually playing, advance.
                    if (EditorApplication.isPlaying)
                    {
                        OnEnteredPlayModeLocked(job);
                        return;
                    }

                    // Never entered play mode within the grace window (e.g. compile error).
                    if (!EditorApplication.isCompiling
                        && !EditorApplication.isPlayingOrWillChangePlaymode
                        && now - job.started_utc_ms > EnterGraceMs)
                    {
                        job.phase = Phase.Failed.ToString();
                        job.reason = "play_mode_never_entered (compile error or blocked play)";
                        job.finished_utc_ms = now;
                        Save(job);
                    }
                    return;
                }

                if (job.phase == Phase.Playing.ToString())
                {
                    long startedPlay = job.play_entered_utc_ms ?? job.started_utc_ms;
                    if (now - startedPlay >= (long)job.duration_seconds * 1000L)
                    {
                        job.phase = Phase.Finalizing.ToString();
                        Save(job);
                        try
                        {
                            if (EditorApplication.isPlaying)
                            {
                                EditorApplication.ExitPlaymode();
                            }
                        }
                        catch (Exception ex)
                        {
                            McpLog.Warn($"[PlaySmokeTest] ExitPlaymode failed: {ex.Message}");
                        }
                    }
                    return;
                }

                if (job.phase == Phase.Finalizing.ToString())
                {
                    // If we somehow returned to edit mode without the state callback,
                    // finalize here as a safety net.
                    if (!EditorApplication.isPlaying)
                    {
                        Finalize(job, failedReason: null);
                    }
                }
            }
        }

        private static void OnEnteredPlayModeLocked(Job job)
        {
            GetCounts(out int e, out int w, out int l);
            job.baseline_errors = e;
            job.baseline_warnings = w;
            job.baseline_logs = l;
            job.rebaselined = true;
            job.play_entered_utc_ms = NowMs();
            job.phase = Phase.Playing.ToString();
            Save(job);
        }

        private static void Finalize(Job job, string failedReason)
        {
            GetCounts(out int errorsNow, out int warningsNow, out int _);

            int errorDelta = Math.Max(0, errorsNow - job.baseline_errors);
            int warningDelta = Math.Max(0, warningsNow - job.baseline_warnings);

            // Error entries the console tracks lump exceptions in with errors; split
            // them out by sampling recent detailed error entries and classifying by
            // signature (reuses ReadConsole's console reflection).
            var sampled = CollectRecentErrors(MaxSampleErrors, out int exceptionCount);

            job.error_count = Math.Max(0, errorDelta - exceptionCount);
            job.exception_count = exceptionCount;
            job.warning_count = warningDelta;
            job.sample_errors = sampled;
            job.finished_utc_ms = NowMs();

            if (failedReason != null)
            {
                job.phase = Phase.Failed.ToString();
                job.reason = failedReason;
            }
            else
            {
                job.phase = Phase.Done.ToString();
            }
            Save(job);
        }

        private static List<string> CollectRecentErrors(int cap, out int exceptionCount)
        {
            exceptionCount = 0;
            var samples = new List<string>();
            try
            {
                var result = ReadConsole.HandleCommand(new JObject
                {
                    ["action"] = "get",
                    ["types"] = new JArray { "error" },
                    ["format"] = "detailed",
                    ["count"] = cap,
                    ["includeStacktrace"] = true,
                });

                var jo = result != null ? JObject.FromObject(result) : null;
                var data = jo?["data"] as JArray;
                if (data == null)
                {
                    return samples;
                }

                foreach (var entry in data)
                {
                    string type = entry["type"]?.ToString() ?? "";
                    string message = entry["message"]?.ToString() ?? "";
                    if (string.Equals(type, "Exception", StringComparison.OrdinalIgnoreCase))
                    {
                        exceptionCount++;
                    }
                    if (samples.Count < cap && !string.IsNullOrEmpty(message))
                    {
                        samples.Add(message);
                    }
                }
            }
            catch (Exception ex)
            {
                McpLog.Warn($"[PlaySmokeTest] Failed to collect recent errors: {ex.Message}");
            }
            return samples;
        }

        // --- SessionState persistence ---

        internal static Job Load()
        {
            try
            {
                string json = SessionState.GetString(SessionKeyJob, string.Empty);
                if (string.IsNullOrWhiteSpace(json))
                {
                    return null;
                }
                return JsonConvert.DeserializeObject<Job>(json);
            }
            catch (Exception ex)
            {
                McpLog.Warn($"[PlaySmokeTest] Failed to load job: {ex.Message}");
                return null;
            }
        }

        internal static void Save(Job job)
        {
            try
            {
                SessionState.SetString(SessionKeyJob, JsonConvert.SerializeObject(job));
            }
            catch (Exception ex)
            {
                McpLog.Warn($"[PlaySmokeTest] Failed to save job: {ex.Message}");
            }
        }

        internal static void ClearPersisted()
        {
            SessionState.EraseString(SessionKeyJob);
        }

        // --- Console counts (reflection, mirrors EditorContext.cs) ---

        private static MethodInfo _getCountsByTypeMethod;
        private static bool _reflectionInitialized;

        private static void EnsureReflection()
        {
            if (_reflectionInitialized)
            {
                return;
            }
            _reflectionInitialized = true;
            try
            {
                Type logEntriesType = typeof(EditorApplication).Assembly.GetType("UnityEditor.LogEntries");
                _getCountsByTypeMethod = logEntriesType?.GetMethod(
                    "GetCountsByType",
                    BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
            }
            catch
            {
                _getCountsByTypeMethod = null;
            }
        }

        private static void GetCounts(out int errors, out int warnings, out int logs)
        {
            errors = warnings = logs = 0;
            EnsureReflection();
            if (_getCountsByTypeMethod == null)
            {
                return;
            }
            try
            {
                var args = new object[] { 0, 0, 0 };
                _getCountsByTypeMethod.Invoke(null, args);
                errors = (int)args[0];
                warnings = (int)args[1];
                logs = (int)args[2];
            }
            catch
            {
                // Leave zeros.
            }
        }

        private static long NowMs() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
    }

    /// <summary>
    /// Re-arms the smoke-test tick and play-mode state callbacks after every domain
    /// reload. Entering play mode reloads the domain, destroying the tick that was
    /// running; this restores it so the job can complete across the reload.
    /// </summary>
    [InitializeOnLoad]
    internal static class PlaySmokeTestReloadHandler
    {
        static PlaySmokeTestReloadHandler()
        {
            EditorApplication.playModeStateChanged += OnPlayModeStateChanged;
            // Only pay the per-frame cost while a job is actually active.
            if (PlaySmokeJobManager.HasActiveJob)
            {
                EditorApplication.update += OnUpdate;
            }
        }

        private static void OnPlayModeStateChanged(PlayModeStateChange change)
        {
            switch (change)
            {
                case PlayModeStateChange.EnteredPlayMode:
                    EnsureTickArmed();
                    PlaySmokeJobManager.OnEnteredPlayMode();
                    break;
                case PlayModeStateChange.EnteredEditMode:
                    EnsureTickArmed();
                    PlaySmokeJobManager.OnEnteredEditMode();
                    break;
            }
        }

        private static void EnsureTickArmed()
        {
            EditorApplication.update -= OnUpdate;
            if (PlaySmokeJobManager.HasActiveJob)
            {
                EditorApplication.update += OnUpdate;
            }
        }

        private static void OnUpdate()
        {
            try
            {
                PlaySmokeJobManager.Tick();
            }
            catch (Exception ex)
            {
                McpLog.Warn($"[PlaySmokeTest] Tick failed: {ex.Message}");
            }

            // Stop ticking once the job is no longer active to avoid idle overhead.
            if (!PlaySmokeJobManager.HasActiveJob)
            {
                EditorApplication.update -= OnUpdate;
            }
        }
    }
}
